src/v/datalake/coordinator/frontend.cc

```diff
@@ -0,0 +1,259 @@
+  , _leaders(leaders)
+  , _shard_table(shards) {}
+
+ss::future<> frontend::stop() { return _gate.close(); }
+
+std::optional<model::partition_id>
+frontend::coordinator_partition(const model::topic_partition& tp) const {
+    const auto md = _metadata->local().get_topic_metadata_ref(
+      model::datalake_coordinator_nt);
+    if (!md) {
+        return std::nullopt;
+    }
+    iobuf temp;
+    write(temp, tp);
+    auto bytes = iobuf_to_bytes(temp);
+    auto partition = murmur2(bytes.data(), bytes.size())
+                     % md->get().get_configuration().partition_count;
+    return model::partition_id{static_cast<int32_t>(partition)};
+}
+
+ss::future<bool> frontend::ensure_topic_exists() {
+    // todo: make these configurable.
+    static constexpr int16_t default_replication_factor = 3;
+    static constexpr int32_t default_coordinator_partitions = 3;
+
+    const auto& metadata = _metadata->local();
+    if (metadata.get_topic_metadata_ref(model::datalake_coordinator_nt)) {
+        co_return true;
+    }
+    auto replication_factor = default_replication_factor;
+    if (replication_factor > static_cast<int16_t>(metadata.node_count())) {
+        replication_factor = 1;
+    }
+
+    cluster::topic_configuration topic{
+      model::datalake_coordinator_nt.ns,
+      model::datalake_coordinator_nt.tp,
+      default_coordinator_partitions,
+      replication_factor};
+
+    topic.properties.compression = model::compression::none;
+    // todo: fix this by implementing on demand raft
+    // snapshots.
+    topic.properties.cleanup_policy_bitflags
+      = model::cleanup_policy_bitflags::none;
+    topic.properties.retention_bytes = tristate<size_t>();
+    topic.properties.retention_local_target_bytes = tristate<size_t>();
+    topic.properties.retention_duration = tristate<std::chrono::milliseconds>();
+    topic.properties.retention_local_target_ms
+      = tristate<std::chrono::milliseconds>();
+
+    try {
+        auto res = co_await _topics_frontend->local().autocreate_topics(
+          {std::move(topic)},
+          config::shard_local_cfg().create_topic_timeout_ms());
+        vassert(
+          res.size() == 1,
+          "Incorrect result when creating {}, expected 1 response, got: {}",
+          model::datalake_coordinator_nt,
+          res.size());
+        if (
+          res[0].ec != cluster::errc::success
+          && res[0].ec != cluster::errc::topic_already_exists) {
+            vlog(
+              datalake::datalake_log.warn,
+              "can not create topic: {} - error: {}",
+              model::datalake_coordinator_nt,
+              cluster::make_error_code(res[0].ec).message());
+            co_return false;
+        }
+        co_return true;
+    } catch (const std::exception_ptr& e) {
+        vlog(
+          datalake::datalake_log.warn,
+          "can not create topic {} - exception: {}",
+          model::datalake_coordinator_nt,
+          e);
+        co_return false;
+    }
```

Comment:
this does not catch generic exceptions like i believe you intended:

https://godbolt.org/z/brPGWxWWo